crates/cargo-gpu/src/spirv_source.rs

```diff
@@ -0,0 +1,264 @@
+//! Use the shader that we're compiling as the default source for which version of `rust-gpu` to use.
+//!
+//! We do this by calling `cargo tree` inside the shader's crate to get the defined `spirv-std`
+//! version. Then with that we `git checkout` the `rust-gpu` repo that corresponds to that version.
+//! From there we can look at the source code to get the required Rust toolchain.
+
+/// The canonical `rust-gpu` URI
+const RUST_GPU_REPO: &str = "https://github.com/Rust-GPU/rust-gpu";
+
+/// The various sources that the `rust-gpu` repo can have.
+/// Most commonly it will simply be the canonical version on crates.io. But it could also be the
+/// Git version, or a fork.
+#[derive(Eq, PartialEq, Clone, Debug)]
+pub enum SpirvSource {
+    /// If the shader specifies a simple version like `spirv-std = "0.9.0"` then the source of
+    /// `rust-gpu` is the conventional crates.io version.
+    ///
+    /// `String` is the simple version like, "0.9.0"
+    CratesIO(String),
+    /// If the shader specifies a version like:
+    ///   `spirv-std = { git = "https://github.com..." ... }`
+    /// then the source of `rust-gpu` is `Git`.
+    ///
+    /// `(String, String)` is the repo source and revision hash or tag.
+    Git((String, String)),
```

Comment:
Do you think we also have to cover the `{ path = "..." }` case? 

And what about when it's mixed like `{ version = "...", path = "..." }` ? 

I initially wanted to model this with a nice type like you have here, but because of the large number of variants I landed on a simple string, which is admittedly not ideal but covers all use cases. What do you think?

Reply:
I've added this, but not tested it yet.